Processor.process() will process any additional Events returned by event handlers

seth-p opened the pull request:
An event handler may return an Event or list (or other iterable) of Events, in which case such event(s) is/are added to a queue of Events to be processed after completion of all processing of the current event.

seth-p committed:
Processor.process() will process events returned by event handlers.

seth-p commented:
This PR breaks the tests, though I must admit I don't understand how the Mock business works.
```
======================================================================
ERROR: test_process_base (tests.unit.test_processor.ProcessorTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\users\seth\github\evproc\.tox\py34\lib\site-packages\mock\mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "C:\Users\Seth\github\evproc\tests\unit\test_processor.py", line 360, in test_process_base
    result = proc.process('ev')
  File "C:\Users\Seth\github\evproc\evproc\processor.py", line 326, in process
    event_deque.extend(proc_result)
TypeError: 'Mock' object is not iterable

======================================================================
ERROR: test_process_stop (tests.unit.test_processor.ProcessorTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\users\seth\github\evproc\.tox\py34\lib\site-packages\mock\mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "C:\Users\Seth\github\evproc\tests\unit\test_processor.py", line 378, in test_process_stop
    result = proc.process('ev')
  File "C:\Users\Seth\github\evproc\evproc\processor.py", line 326, in process
    event_deque.extend(proc_result)
TypeError: 'Mock' object is not iterable

----------------------------------------------------------------------
Ran 59 tests in 0.162s

FAILED (errors=2)
ERROR: InvocationError: 'C:\\Users\\Seth\\github\\evproc\\.tox\\py34\\Scripts\\nosetests.EXE -v'
```

klmitch commented on evproc/processor.py:
```diff
@@ -15,7 +15,8 @@
 
-from evproc import exc
+from collections import deque
+from evproc import exc, Event
```
I prefer importing modules and _not_ objects, so turn these into:

```
import collections
import evproc
from evproc import exc
```

klmitch commented on evproc/processor.py:
```diff
@@ -308,11 +310,19 @@ def process(self, ev):
-                return ex.retval
+        event_deque = deque()
+        event_deque.append(ev)
+        while len(event_deque):
```
Shouldn't need the len() here; a simple "while event_deque" should do the trick.

klmitch commented on evproc/processor.py:
```diff
@@ -308,11 +310,19 @@ def process(self, ev):
+                    return ex.retval
+                if isinstance(proc_result, Event):
+                    event_deque.append(proc_result)
+                elif proc_result is not None:
```
If you're importing collections, you could use isinstance() with collections.Sequence.  That would be true for strings as well, though, so you have to deal with that.

klmitch commented:
Well, I'll need to see the tests pass.  I would also like to see tests added to cover the additional branches.

You may want to refer to [1] for documentation on the mock module.  One of the operative points is that, unless you override the return_value, another instance of mock.Mock will be returned, which is why you're seeing the test failures you're seeing.  You can set return_value for the existing processor mocks to None, which will test the back-compatibility branch of your changes, but as I say, I'd like to see additional tests as well to cover the additional branches.

You may want to know about how to measure test coverage; I do that with "tox -e cover", and I typically shoot for 100% coverage.  This change is simple enough that I'd like to see the tests updated to maintain that level of coverage.  (I actually typically develop with reference to "tox -e cover,pep8", so I also see the style test results.  Also, I like to "rm .coverage" if I've made any changes to the main code since the last time I ran "tox -e cover", so that coverage will be fully recomputed.)

[1] http://www.voidspace.org.uk/python/mock/

seth-p commented:
OK, I think I've addressed all the issues other than the tests. I simplified the code slightly by assuming that the return value from an event handler, if not None, must be an iterable.

I will try to fix the tests soon.

seth-p commented:
OK, I've updated the tests, and everything passes (with complete coverage).

seth-p commented:
@klmitch, does this look good? Any further comments?

klmitch commented on evproc/processor.py:
```diff
@@ -13,6 +13,7 @@
 #    governing permissions and limitations under the License.
 
+import collections
 import stevedore
```
Put a blank line after the "import collections"; I try to conform to OpenStack style, which has 3 blocks of imports, separated by blank lines: standard library imports (e.g., collections); third-party imports (e.g., stevedore); and internal imports (e.g., evproc.exc).

klmitch commented:
Sorry for the delay in response; I was a bit sidetracked this weekend.  This is looking pretty good, aside from the one style comment I noted inline.

One question, though: the application I originally wrote this for actually calls `proc.process(event)` on several events at the end of a processor.  Could you contrast that approach to this approach?  Why would your approach work better in this case?

seth-p commented:
I added the blank line after "import collections."

If a handler for event1 calls proc.process(event2), then event2 is processed immediately, regardless of whether there are additional handlers for event1 that haven't yet run. That may be desired, or it may not. If one first wants all the handlers for event1 to run before processing event2, then, absent this PR, one would have to make sure to call proc.process(event2) from the final event1 handler, which may not always be the natural place to decide that one needs to process event2.

klmitch commented on .gitignore:
```diff
@@ -1,5 +1,6 @@
 /.DS_Store
 /.coverage
+/.idea
```
extraneous for me, but no big deal…

seth-p replied:
This is from PyCharm...

klmitch commented on evproc/processor.py:
```diff
@@ -308,11 +311,16 @@ def process(self, ev):
+                try:
+                    additional_events = proc(ev)
+                except exc.StopProcessing as ex:
+                    return ex.retval
```
Let's say you have two processors set up for event1, call them proc1 and proc2 (in that order).  Let's say proc1 fires off event2 and event3 by returning them as a list, and proc2 then raises StopProcessing.  The return here means that event2 and event3 will never be processed.  Is that what we should go for here?  I think the primary reasoning that I chose not to do the additional_events-type processing is because of that case, and I felt that this kind of architecture would be best served by an asynchronous event loop; evproc I designed as a synchronous event loop.  What do you think?

seth-p replied:
Good observation. I hadn't thought of this case. I think this behavior (i.e. event2 and event3 aren't processed) is reasonable.

An alternative would be to stop calling any additional event1 handlers (if such exist), but to continue processing the event2 and event3 returned by proc1. This seems like a perfectly plausible alternative, except that in this version I don't know what should be returned if multiple processors raise StopProcessing (e.g. one for event1, one for event2, and one for event3). If we really wanted to handle it this way, I suppose we could "yield ex.retval" rather than "return ex.retval" each time a handler raises StopProcessing, though this may be a bit odd (though has the advantage of occurring at the right time, e.g. we'd yield the ex.retval from proc2 before processing event2 and event3).

seth-p replied:
With regard to the alternative behavior described above, one could support two versions of StopProcessing: one that stops processing the current event, but proceeds to process any additional events in the queue; and one that stops processing and immediately returns, ignoring any remaining events in the queue.

klmitch replied:
Hmmm…I like the split on StopProcessing approach.  It could even be done using the existing StopProcessing; just add an additional argument to signal which behavior to use, and give it a reasonable default.

seth-p replied:
OK, I'll try to come up with something.

In the case of (multiple) stop-with-this-event-but-proceed-with-remaining-events StopProcessing exceptions, what should process() return? Should it yield the retval from each one? Return a list containing all of them?

klmitch replied:
Let's not change this into a generator, as that requires a backwards-incompatible change; that means we have to do a return.  I say we should only do a return if we're stopping all event processing.

In fact, I'm tempted to suggest that we should only return something if the initial event raises a StopProcessing with the appropriate settings, and ignore return code in any other StopProcessing exception.  The mechanism is intended to not just allow event processing to be stopped, but to also potentially allow a value for the event processing to be returned to whatever launched the event in the first place; in that case, an alternative event might be an entirely different interface.

seth-p replied:
The simplest way to implement this, I think, is that if a StopProcessing exception contains a retval (which may be None), then Processor.process() will immediately return that retval and ignore any remaining events in the queue; if a StopProcessing exception does not contain a retval, then Processor.process() will stop processing the current event (ignoring any remaining handlers for that event) and proceed to process the next event in the queue. This would be independent of whether or not the event is the first one.

I think this would be backwards compatible, and would be rather intuitive, I think.

I just updated the commit to implement this, though I haven't yet implemented tests. Let me know what you think.

klmitch replied:
I'm not convinced.  The original behavior was, if `StopProcessing` is raised, return whatever the `retval` is…and `retval` defaults to `None`.  The code I have that uses evproc in fact does a `raise evproc.StopProcessing()` and expects all event processing to halt.  Of course, those processors aren't returning any additional events, as they would use `proc.process()` to accomplish the same thing…

seth-p replied:
If there are no other events in the queue (which there almost certainly won't be with existing code), then the behavior is unchanged, whether or not `retval` is explicitly set in `StopProcessing`.

Anyway, up to you.

seth-p replied:
My bad. I see that `StopProcessing.__init__()` takes `retval=None`, so that `retval` will always be set. So I guess for my code to work right I'd need to modify `StopProcessing.__init__()` to set `retval` only if it's explicitly provided. This would still be backwards compatible in the case of no additional events in the queue.

klmitch replied:
Yeah, I knew about the default value for `retval`, and was going to mention the idiom I usually use for that sort of thing (`_unset = object()` and use `_unset` as the default), but I wanted to see what sort of defense you could come up with :)

I think you have a good point that the behavior may be unchanged, but don't forget that this is pluggable; you could have some event processors for one event that are coded to the old version, while others for that same event are coded to the new version.  Still, the new ones could probably be written to deal with that case, but it should probably be called out in the documentation…

seth-p commented:
I've updated the code so that raising a `StopProcessing(retval)` will cause all processing to stop (ignoring any additional events in the queue) and returning `retval` (which may be `None`); whereas raising `StopProcessing()` (with no `retval` specified) will cause any additional handlers for the current event to be skipped, but processing of remaining events in the queue will continue.

If this looks ok, I will update the documentation.

klmitch commented on evproc/exc.py:
```diff
@@ -57,4 +60,5 @@ def __init__(self, retval=None):
 
-        self.retval = retval
+        if retval is not _unset:
+            self.retval = retval
```
I'd prefer setting `retval` unconditionally and then picking the behavior based on whether its value is `_unset` or not…

klmitch commented:
The basic approach is satisfactory; have just one little tweak pointed out inline, then of course the doc update.  Thanks for bearing with me :)

seth-p commented:
I've updated the code and README.rst

klmitch commented on evproc/exc.py:
```diff
@@ -47,7 +47,9 @@ class StopProcessing(Exception):
     """
 
-    def __init__(self, retval=None):
+    _unset = object()
```
I would prefer this to be a top-level constant, as opposed to a class constant.

seth-p replied:
I  changed _unset to a top-level constant.

klmitch commented:
Looks good.  One nit noted inline.  Note: it may be tomorrow before I get time to merge this; I want to run the tests myself before I merge.  Thanks for your patience!

klmitch commented:
Man, I've been forgetting about this or too busy to deal with this!  Sorry for the delay :/  'fraid it's going to be a bit longer, but I promise I will try to get to this.

seth-p commented:
Any further thoughts?

klmitch commented:
OK, this has sat here long enough; I apologize for the length of time it has taken for me to do my final review, but it is now done, and I approve…

klmitch commented:
And FYI, I just pushed evproc 0.2.0 to PyPi…

seth-p commented:
Thanks.

By the way, is there any particular reason that Python 3.5 isn't supported?

klmitch commented:
I just don't explicitly call it out; it should work, based on the fact that it works on Python 3.4, but I haven't yet installed 3.5 in the local directory I keep alternate versions of Python installed in, so I can't yet run the tests in that environment to double-check.  I'll get around to doing that one of these days :)
